src/Admin/Controllers/ToolsController.cs

```diff
@@ -313,6 +315,71 @@ public async Task<IActionResult> TaxRateAddEdit(string stripeTaxRateId = null)
             return View(model);
         }
 
+        public async Task<IActionResult> TaxRateUpload(IFormFile file)
+        {
+            if (file == null || file.Length == 0)
+            {
+                throw new ArgumentNullException(nameof(file));
+            }
+
+            // Build rates and validate them first before updating DB & Stripe
+            var taxRateUpdates = new List<TaxRate>();
+            var currentTaxRates = await _taxRateRepository.GetAllActiveAsync();
+            using var reader = new StreamReader(file.OpenReadStream());
+            while (!reader.EndOfStream)
+            {
+                var line = await reader.ReadLineAsync();
+                if (string.IsNullOrWhiteSpace(line))
+                {
+                    continue;
+                }
+                var taxParts = line.Split(',');
+                if (taxParts.Length < 2)
```

Comment:
normally i would caution against parsing CSV and instead use a library like CSVHelper, but I guess this seems simple enough.